src/rust/core/server/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+//!                         └ ─ ─ ─ ─ ─
+//! ```
+//!
+//! Multiple worker thread model:
+//! ```text
+//! ┌──────────┐                 ┌──────────┐
+//! │  admin   │                 │ listener │
+//! │          │                 │          │
+//! │  :9999   │                 │  :12321  │
+//! └──────────┘                 └──────────┘
+//!                                    │
+//!                            ┌───────┴───────┐
+//!                            │               │
+//!                            ▼               ▼
+//! ┌ ─ ─ ─ ─ ─          ┌──────────┐    ┌──────────┐        ┌ ─ ─ ─ ─ ─
+//!            │─        │          │    │          │                   │─
+//! │  client   ◀┼──────▶│  worker  │    │  worker  │◀──────▶│  client    │─
+//!            │   │     │          │    │          │                   │   │
+//! └ ─ ─ ─ ─ ─  │       └──────────┘    └──────────┘        └ ─ ─ ─ ─ ─  │
+//!   └ ─ ─ ─ ─ ─  │           ▲               ▲               └ ─ ─ ─ ─ ─  │
+//!     └ ─ ─ ─ ─ ─            └───────┬───────┘                 └ ─ ─ ─ ─ ─
+//!                                    ▼
+//!                              ┌──────────┐
+//!                              │          │
+//!                              │ storage  │
+//!                              │          │
+//!                              └──────────┘
+//! ```
+//!
+//! ## Control Plane
+//! The control plane is handled by a single `admin` thread. This thread is
+//! responsible for handling administrative commands and metrics exposition.
+//!
+//! ## Data Plane
+//! The data plane is handled by two or more threads depending on the
+//! configuration.
+//!
+//! ### Listener
+//! At a minimum we have one `listener` thread which owns the listening socket,
+//! accepts new connections, and handles TLS negotiation if it is enabled by the
+//! configuration. Fully negotiated sessions are then handed off to one or more
+//! worker threads.
+//!
+//! ### Worker
+//! Worker threads handle ongoing communications for an established session.
+//! This includes request parsing and response composition. If the configuration
+//! specifies a single worker thread, the thread also owns the underlying cache
+//! datastructure. However, if the configuration specifies multiple worker
+//! threads a fully parsed request is handed over to the `storage` thread for
+//! execution.
+//!
+//! ### Storage
+//! An optional thread which is used only if there is more than one worker
+//! thread configured. This thread is used to own the cache datastructure and
+//! execute requests. The storage thread will receive requests from a worker
+//! over a queue, execute the request, and returns the result back to the worker
+//! thread.
+
+#[macro_use]
+extern crate rustcommon_logger;
+
+#[macro_use]
+extern crate rustcommon_fastmetrics;
+
+mod poll;
+mod process;
+mod threads;
+
+pub use process::{Process, ProcessBuilder};
+
+const DEFAULT_BUFFER_SIZE: usize = 1024; // 1KB
```

Comment:
For session buffer, I argue this default is too low for the following reasons:
1. The default should at least be bigger than 1 MTU;
2. The default should ideally be large enough that resizing is uncommon for most workloads (for Twitter, that's actually pretty small, but I think p95 is still greater than 1KB);
3. For TLS, since there's an obvious upper bound on fragment size, which is `2^14=16 KiB` per RFC 5246 ([ref](https://datatracker.ietf.org/doc/html/rfc5246#section-6.2.1)), so I say we just set the default size to that. This may mean we set the same for plaintext for simplicity.

Reply:
Yeah, this is definitely low. I like the idea of adjusting to 16KiB to match the TLS fragment size. See comment above for thoughts on resize strategy.

Reply:
Bumped to 16KiB and added a comment w/ link to the RFC